Add `SortingAnalyzer.set_sorting_property()/get_sorting_property()` functions

alejoe91 opened the pull request:
Helper function to add properties to `SortingAnalyzer.sorting` and make them persistent to disk

zm711 commented on src/spikeinterface/core/sortinganalyzer.py:
```diff
@@ -742,6 +742,53 @@ def set_temporary_recording(self, recording: BaseRecording, check_dtype: bool =
+        Set property vector for unit ids.
+
+        If the SortingAnalyzer backend is in memory, the property will be only set in memory.
+        If the SortingAnalyzer backend is in binary_folder or zarr, the property will also
```
```suggestion
        If the SortingAnalyzer backend is `binary_folder` or `zarr`, the property will also
```

zm711 commented on src/spikeinterface/core/sortinganalyzer.py:
```diff
@@ -742,6 +742,53 @@ def set_temporary_recording(self, recording: BaseRecording, check_dtype: bool =
+        values : np.array
+            Array of values for the property
+        ids : list/np.array, default: None
+            List of subset of ids to set the values, default: None
```
```suggestion
            List of subset of ids to set the values.
```

zm711 commented on src/spikeinterface/core/sortinganalyzer.py:
```diff
@@ -742,6 +742,53 @@ def set_temporary_recording(self, recording: BaseRecording, check_dtype: bool =
+            Array of values for the property
+        ids : list/np.array, default: None
+            List of subset of ids to set the values, default: None
+            if None which is the default all the ids are set or changed
```
```suggestion
            if None all the ids are set or changed
```
that's definitely an an assisted docstring ;)

zm711 commented on src/spikeinterface/core/sortinganalyzer.py:
```diff
@@ -742,6 +742,53 @@ def set_temporary_recording(self, recording: BaseRecording, check_dtype: bool =
+            if None which is the default all the ids are set or changed
+        missing_value : object, default: None
+            In case the property is set on a subset of values ("ids" not None),
+            it specifies the how the missing values should be filled.
+            The missing_value has to be specified for types int and unsigned int.
```
```suggestion
           This argument specifies how to fill missing values
            The `missing_value` has to be specified for types int and unsigned int.
```
Is this true? It has to be int?

zm711 commented on src/spikeinterface/core/sortinganalyzer.py:
```diff
@@ -742,6 +742,53 @@ def set_temporary_recording(self, recording: BaseRecording, check_dtype: bool =
+                zarr_root = self._get_zarr_root(mode="r+")
+                prop_values = self.sorting.get_property(key)
+                if prop_values.dtype.kind == "O":
+                    warnings.warn(f"Property {key} not saved because it is a python Object type")
```
Can we tell the user how to fix it or is this just a you're out of luck?

alejoe91 replied:
I think this is by construction, so we can re-open the zarr folder also outside of Python

zm711 commented on src/spikeinterface/core/tests/test_sortinganalyzer.py:
```diff
@@ -66,6 +66,12 @@ def test_SortingAnalyzer_memory(tmp_path, dataset):
     assert not sorting_analyzer.return_scaled
 
+    # test set_unit_property
+    sorting_analyzer.set_unit_property(key="quality", values=["good"] * len(sorting_analyzer.unit_ids))
```
because values is a string here so why does missing value have to be an int?

alejoe91 replied:
it doesn't have to be an int. It is required if the values are integers. I changed the docstrings to make it more clear

zm711 reviewed (commented):
can we get a test for the missing_value argument?

samuelgarcia commented on src/spikeinterface/core/sortinganalyzer.py:
```diff
@@ -742,6 +742,53 @@ def set_temporary_recording(self, recording: BaseRecording, check_dtype: bool =
+        key,
+        values: list | np.ndarray | tuple,
+        ids: list | np.ndarray | tuple | None = None,
+        missing_value: Any = None,
```
I would explicitly make a `save=True/False`, option here.

alejoe91 commented on src/spikeinterface/core/sortinganalyzer.py:
```diff
@@ -742,6 +742,53 @@ def set_temporary_recording(self, recording: BaseRecording, check_dtype: bool =
+        ids : list/np.array, default: None
+            List of subset of ids to set the values, default: None
+            if None which is the default all the ids are set or changed
+        missing_value : object, default: None
```
```suggestion
        missing_value : Any, default: None
```

chrishalcrow commented:
Hello, found a bug:

``` python
import spikeinterface.full as si
rec, sort = si.generate_ground_truth_recording()
sa = si.create_sorting_analyzer(recording=rec, sorting=sort)
sa.set_unit_property(key="quality", values=["bad"]*10)
sa.set_unit_property(key="quality", values=["good","good","good"], ids=['1','7','8'])
print(sa.get_sorting_property("quality"))
>>> array(['bad', 'goo', 'bad', 'bad', 'bad', 'bad', 'bad', 'goo', 'goo',
       'bad'], dtype='<U3')
```

It's saving the labels as `U3` then using this later, so that only the first three chracters of `'good'` are saved.

chrishalcrow commented on src/spikeinterface/core/sortinganalyzer.py:
```diff
@@ -742,6 +743,56 @@ def set_temporary_recording(self, recording: BaseRecording, check_dtype: bool =
             warnings.warn("SortingAnalyzer recording is already set. The current recording is temporarily replaced.")
         self._temporary_recording = recording
 
+    def set_unit_property(
```
Should we use `set_sorting_property` to match `get_sorting_property`? I prefer get/set_unit_... but we've already used sorting.

alejoe91 replied:
makes sense :)

alejoe91 replied:
done

chrishalcrow commented:
The error message when you put in a non_existent id mentions `channel_ids`, which might be confusing:

e.g. forgetting that generated ids are strings...
``` python
sa.set_unit_property(key="beep", values=[5,4,3,2,1], ids=[1,2,3,4,5])
>>> ValueError: IDs [np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5)] are not channel ids of the extractor.
Available ids are ['0' '1' '2' '3' '4' '5' '6' '7' '8' '9'] with dtype <U1
```

(The fact this dtype is <U1 looks a big dangerous!!)

Error message here:
https://github.com/SpikeInterface/spikeinterface/blob/91c2a0d8f57bd5b8ef999e1dc9452643acac4f83/src/spikeinterface/core/base.py#L148

alejoe91 commented:
@chrishalcrow I fixed the unicode size issue in the core `get_property` by resizing the dtype according to existing and new values.

Also renamed to `set_sorting_property`, added `get_sorting_property`, and removed reference to "channel IDs" in the error message.

Good to go on my end!

chrishalcrow reviewed (approved):
Thanks Alessio - really helpful feature! This
``` python
sa.select_units(sa.unit_ids[sa.get_sorting_property('brain_area')=='MEC'])
```
feels good!!

All working for me :)

zm711 reviewed (approved):
Thanks the docs are clearer now!

alejoe91 commented:
Also fixes #3699 @chrishalcrow

samuelgarcia commented on src/spikeinterface/core/sortinganalyzer.py:
```diff
@@ -1186,11 +1259,16 @@ def get_sorting_provenance(self):
+                # try-except here is because it's not required to be able
+                # to load the sorting provenance, as the user might have deleted
+                # the original sorting folder
+                try:
+                    sort_dict = zarr_root["sorting_provenance"][0]
+                    sorting_provenance = load(sort_dict, base_folder=self.folder)
+                except:
+                    pass
```
I think we should raise an error when the user really wants to load this no ?
I am not sure to like theses try except at many place when dealing with zarr

samuelgarcia commented:
Nobody is giving me a chance to read teh PR before the merge now...

zm711 commented:
Is it finally time for us to have a ready-to-merge tag?

alejoe91 commented:
> Is it finally time for us to have a ready-to-merge tag?

maybe! but this is just @samuelgarcia with his daily dose of complaints :)

zm711 commented on src/spikeinterface/core/base.py:
```diff
@@ -293,6 +293,11 @@ def set_property(
                     ), f"Mismatch between existing property dtype {existing_property.kind} and provided values dtype {dtype_kind}."
 
                 indices = self.ids_to_indices(ids)
+                if dtype_kind == "U":
```
So your idea here would be no matter what annotation as long as it is a string it can be combined and then you'll resize? I guess that makes sense right. All strings ought to be combinable? but something like a str and an int shouldn't be ?

chrishalcrow replied:
Another thing: numpy allows for concatenation of different unicode types:
```np.concat([np.array(['hello', 'there']),  np.array(['you'])])```
is fine. So we don't necessarily need to update the data ourselves - we just need to check that it's all unicode.
